### ISSUE TITLE
kernel warning

### ISSUE THREAD

**Issue by HI0T** [intel/QAT_Engine#18]
Hi.
I build QAT1.6 driver and QAT Engine 0.5.21.
When trying to run ./openssl speed -engine qat -elapsed -async_jobs 72 rsa2048
syslog displays these warnings:
```Apr 24 11:45:37 kernel: [   92.235023] ------------[ cut here ]------------
Apr 24 11:45:37 kernel: [   92.235029] WARNING: CPU: 3 PID: 2516 at /build/linux-wXdoVv/linux-4.4.0/mm/mmap.c:1640 mmap_region+0x55b/0x650()
Apr 24 11:45:37 kernel: [   92.235030] Modules linked in: qat_contig_mem(OE) ipt_MASQUERADE nf_nat_masquerade_ipv4 xfrm_user xfrm_algo iptable_nat nf_conntrack_ipv4 nf_defrag_ipv4 nf_nat_ipv4 xt_addrtype iptable_filter ip_tables xt_conntrack x_tables nf_nat nf_conntrack br_netfilter bridge stp llc aufs pci_stub vboxpci(OE) vboxnetadp(OE) vboxnetflt(OE) vboxdrv(OE) icp_qa_al(OE) zlib joydev input_leds snd_usb_audio eeepc_wmi asus_wmi sparse_keymap hid_plantronics hid_generic snd_usbmidi_lib intel_rapl x86_pkg_temp_thermal intel_powerclamp coretemp usbhid hid kvm_intel snd_hda_codec_hdmi snd_hda_codec_realtek snd_hda_codec_generic kvm snd_hda_intel irqbypass snd_hda_codec binfmt_misc snd_hda_core crct10dif_pclmul crc32_pclmul snd_hwdep ghash_clmulni_intel aesni_intel aes_x86_64 lrw gf128mul glue_helper ablk_helper snd_pcm cryptd serio_raw lpc_ich tpm_infineon 8250_fintek snd_seq_midi snd_seq_midi_event snd_rawmidi mei_me snd_seq snd_seq_device snd_timer snd soundcore mei shpchp acpi_pad mac_hid parport_pc ppdev lp parport autofs4 mxm_wmi i915 i2c_algo_bit drm_kms_helper psmouse syscopyarea sysfillrect sysimgblt ahci fb_sys_fops libahci drm r8169 mii video wmi fjes
Apr 24 11:45:37 kernel: [   92.235070] CPU: 3 PID: 2516 Comm: openssl Tainted: G           OE   4.4.0-72-generic #93-Ubuntu
Apr 24 11:45:37 kernel: [   92.235071] Hardware name: ASUS All Series/Z97-K, BIOS 2604 05/20/2015
Apr 24 11:45:37 kernel: [   92.235072]  0000000000000286 000000004507e18e ffff8803c491bd48 ffffffff813f82b3
Apr 24 11:45:37 kernel: [   92.235074]  0000000000000000 ffffffff81cd2908 ffff8803c491bd80 ffffffff81081302
Apr 24 11:45:37 kernel: [   92.235075]  ffff8800c9d73be8 00007faad8ca0000 0000000000000000 ffff8800c6d47d48
Apr 24 11:45:37 kernel: [   92.235077] Call Trace:
Apr 24 11:45:37 kernel: [   92.235081]  [<ffffffff813f82b3>] dump_stack+0x63/0x90
Apr 24 11:45:37 kernel: [   92.235083]  [<ffffffff81081302>] warn_slowpath_common+0x82/0xc0
Apr 24 11:45:37 kernel: [   92.235085]  [<ffffffff8108144a>] warn_slowpath_null+0x1a/0x20
Apr 24 11:45:37 kernel: [   92.235086]  [<ffffffff811c956b>] mmap_region+0x55b/0x650
Apr 24 11:45:37 kernel: [   92.235088]  [<ffffffff811c9993>] do_mmap+0x333/0x420
Apr 24 11:45:37 kernel: [   92.235090]  [<ffffffff811adebf>] vm_mmap_pgoff+0xaf/0xe0
Apr 24 11:45:37 kernel: [   92.235092]  [<ffffffff811c7d81>] SyS_mmap_pgoff+0x1c1/0x290
Apr 24 11:45:37 kernel: [   92.235096]  [<ffffffff8103406b>] SyS_mmap+0x1b/0x30
Apr 24 11:45:37 kernel: [   92.235099]  [<ffffffff8183c672>] entry_SYSCALL_64_fastpath+0x16/0x71
Apr 24 11:45:37 kernel: [   92.235100] ---[ end trace c915859e9ebb5592 ]---
```
what does that mean?

**Comment by stevelinsell** [comment]
Hi,

The crash is happening in the kernelspace part of the qat_contig_mem memory driver when it tries to mmap a region of memory into userspace. 
I'm not sure what is causing this.
Can you confirm whether you are running as root user as it maybe a permissions problem?
Could you also try running the test application that builds in the same directory as the qat_contig_mem. It is called qat_contig_mem_test and will just try and allocate and mmap into userspace a small chunk of memory to test out the qat_contig_mem driver. It would be useful to know if you see the same issue when running that.

Kind Regards,

Steve.

**Comment by HI0T** [comment]
Under the root the same error
test passes:
```
seg mapped to 0x7fad3f18b000, virtualAddress in seg 0xffff880348d1e000, length 64
Hello world!
```

**Comment by stevelinsell** [comment]
Ok I've investigated this.
The output you are seeing is just a warning and only occurs the first time you attempt to do a mmap.
It is caused because the qat_contig_mem uses a technique to align the mmapped region on a virtual address boundary that is a multiple of the size of the slab being allocated. It does this because the slab has a header that contains important information such as the physical address of the slab. By mapping the region on a multiple of the size of slab allows the header to be found purely using maths (as long as you know the size of the slab (we use a fixed size slab so it never changes). This allows faster virtual to physical address lookups. This is done by specifying a larger region to mmap and then adjusting the addresses.
You'll only see the error on the first slab allocation of using the qat_contig_mem driver. All subsequent allocations will not flag the error as the kernel is using a WARN_ON_ONCE macro, see here:
http://lxr.free-electrons.com/source/mm/mmap.c?v=4.4#L1636

According to comments in the kernel other drivers also change the vma->vm_start address.
I believe this warning is only flagged the once just to make people aware that start address has been changed in case that is unexpected. It is safe to ignore this warning and it should not appear again until the qat_contig_mem driver is reloaded and the first slab is allocated.
